core/parentchain/block-import-dispatcher/src/immediate_dispatcher.rs

```diff
@@ -17,16 +17,28 @@
 
 use crate::{error::Result, DispatchBlockImport};
 use itc_parentchain_block_importer::ImportParentchainBlocks;
-use std::{sync::Arc, vec::Vec};
+use log::*;
+use std::{boxed::Box, vec::Vec};
 
 /// Block import dispatcher that immediately imports the blocks, without any processing or queueing.
 pub struct ImmediateDispatcher<BlockImporter> {
-	block_importer: Arc<BlockImporter>,
+	block_importer: BlockImporter,
+	import_event_observers: Vec<Box<dyn Fn() + Send + Sync + 'static>>,
 }
 
 impl<BlockImporter> ImmediateDispatcher<BlockImporter> {
-	pub fn new(block_importer: Arc<BlockImporter>) -> Self {
-		ImmediateDispatcher { block_importer }
+	pub fn new(block_importer: BlockImporter) -> Self {
+		ImmediateDispatcher { block_importer, import_event_observers: Vec::new() }
+	}
+
+	pub fn with_observer<F>(self, callback: F) -> Self
+	where
+		F: Fn() + Send + Sync + 'static,
+	{
+		let mut updated_observers = self.import_event_observers;
+		updated_observers.push(Box::new(callback));
+
+		Self { block_importer: self.block_importer, import_event_observers: updated_observers }
```

Comment:
I have looked at what the streams trait, and it seems to be an async, multi-threaded pattern, like the `mpsc` crate with `channel`. In this case we only have a single thread (so simple way to make it multi-threaded), and everything happens synchronously with a simple call-back. And on top of that, I'm uncertain about its availability in `sgx`, since it lives in the `futures` crate, of which there is an `sgx` port, but not with full features I believe.

Reply:
I see that makes total sense, thanks! The only other thing I could imagine is something built on channel by ourselves:

```rust

pub struct ImmediateDispatcher<BlockImporter> {
	block_importer: BlockImporter,
	import_event_notifier: Vec<Sender<ImportNotification>>,
}

pub enum ImportNotification {
     BlockImportSuccess,
     BlockImportFailed,
}

...
fn dispatch_import(&self, blocks: Vec<Self::SignedBlockType>) -> Result<()> {
      self.block_importer.import_parentchain_blocks(blocks).map_err(|e| e.into())
      self.block_importer.import_parentchain_blocks(blocks)?;
      self.import_event_observers.iter().for_each(|sender| sender.send(ImportNotification::BlockImportSuccess);
      Ok(())
}
```

And after I wrote above snippet, I had to realize that this is not possible because we are single-threaded. XD

I thought I will still leave it here, just for inspiration.

Reply:
Yes, in a synchronous single-threaded environment, this above is basically the same as what we have now 😄